include/rocksdb/cache.h

```diff
@@ -273,11 +281,27 @@ class Cache {
       // default implementation is noop
   }
 
-  // Apply callback to all entries in the cache
-  // If thread_safe is true, it will also lock the accesses. Otherwise, it will
-  // access the cache without the lock held
-  virtual void ApplyToAllCacheEntries(void (*callback)(void*, size_t),
-                                      bool thread_safe) = 0;
+  // Apply a callback to all entries in the cache. The Cache must ensure
+  // thread safety but does not guarantee that a consistent snapshot of all
+  // entries is iterated over if other threads are operating on the Cache
+  // also. If the Cache uses locks, setting `average_entries_per_lock` to
+  // a higher value suggests iterating over more entries each time a lock
+  // is acquired, likely reducing the time for ApplyToAllEntries but
+  // increasing latency for concurrent users of the Cache. Setting
+  // `average_entries_per_lock` to a smaller value could be helpful if
+  // callback is relatively expensive, such as using large data structures.
+  virtual void ApplyToAllEntries(
+      const std::function<void(const Slice& key, void* value, size_t charge,
+                               DeleterFn deleter)>& callback,
+      size_t average_entries_per_lock = 256) = 0;
```

Comment:
Is this an incompatible change and is there any way around it?  As it stands now, any existing Cache implementation will not work/compile.  

Are we okay with forcing code updates?

Reply:
I think custom Cache implementations are rare enough, and the likely benefit high enough to impose this. I tried other approaches to gathering the statistics but they hit critical areas of performance. Benchmark testing on this approach looks like it will be almost free even if we gathered the stats at intervals of seconds rather than minutes or hours.

Rules (e.g. "no breaking API changes") are negotiable, especially at Facebook. ;)

Reply:
Nit: The google coding standard does not allow default arguments on virtual functions.  Can we split this into two methods, one (not virtual) with no "average_entries_per_lock" and a virtual one without the default